classroom_server_api/helper_funcs.R

```diff
@@ -0,0 +1,202 @@
+  id_str <- make_id_str(...)
+  query <- glue::glue(
+    "SELECT COUNT(*)
+      FROM {class_table(what)}
+      WHERE {id_str};"
+  )
+  
+  n <- dbGetQuery(con, query)
+  
+  if (n == 0) stop("Invalid IDs, no objects found.")
+  if (n > 1) stop("Multiple objects with that ID set.")
+  
+  NULL
+}
+
+# create object in table
+create_object <- function(which, ...) {
+  fields <- names(list(...)) %>% paste(collapse = ", ")
+  values <- c(...) %>% safe_chr() %>% paste(collapse = ", ")
+  
+  query <- glue::glue(
+    "INSERT INTO {class_table(which)}
+            ({fields})
+            VALUES ({values})
+            RETURNING *
+            ;"
+  )
+  
+  print(query)
+  
+  dbGetQuery(con, query)
+}
+
+# Identify and get attributes ------
+
+# Check that attribute is valid for type of object
+attr_valid <- function(which, attr, editable_only = FALSE) {
+  if(!attr %in% attributes(which, editable_only)) {
+    stop(glue::glue("Valid attributes are {attributes(which, editable_only)}, not {attr}."))
+  }
+  NULL
+}
+
+# Check that calling attribute of particular object is  valid
+attr_exists <-function(which , attr, ..., editable_only = FALSE) {
+  exists(which, ...)
+  attr_valid(which, attr, editable_only)
+}
+
+# Get attribute from a particular object; ... should be named ids
+get_attr <- function(which, attr, ...) {
+  id_str <- make_id_str(...)
+  query <- glue::glue(
+    "SELECT {attr}
+      FROM {class_table(which)}
+      WHERE {id_str};"
+  )
+  
+  dbGetQuery(con, query)
+}
+
+# Update attribute for particular object; ... should be named ids
+set_attr <- function(which, attr, value, ...) {
+  attr_exists(which, attr, editable_only = TRUE, ...)
+  value <- safe_chr(value)
+  
+  query <- glue::glue(
+    "UPDATE {class_table(which)}
+      SET {attr} = {value}
+      WHERE {make_id_str(...)}
+    RETURNING *;"
+  )
+  
+  dbGetQuery(con, query)
+}
+
+
+# gettable/editable attributes for each object type
+attributes <- function(which, editable_only = FALSE) {
+  edit_attrs <- list(
+    classroom = c("name", "description", "password", "status"), 
+    instance = c("identifier", "url", "username", "password"), 
+    student = c("name", "email", "consent", "cookie", "other"), 
+    event = c("event", "session", "cookie", "other")
+  )
+  
+  static_attrs <- list(
+    classroom = c("id"), 
+    instance = c("id", "classroomid"), 
+    student = c("id", "classroomid"), 
+    event = c("id", "classroomid", "instanceid", "studentid")
+  )
+  
+  attrs <- do.call(switch, c(EXPR = which, edit_attrs))
+  
+  if (!editable_only) attrs <- c(attrs, 
+                                 do.call(switch, c(EXPR = which, static_attrs)))
+  
+  attrs
+}
+
+# Specific helpers -----
+
+is_claimed <- function(instance_id) {
+  instance_id %in% get("claim")$instanceid
+}
+
+archive_class <- function(class_id) {
+  query <- glue::glue(
+    "UPDATE {class_table('classroom')}
+      SET 
+        name = concat('ARCHIVED {Sys.Date()}', name),
```

Comment:
This was probably in the specs I gave you, but I'm wondering now if we should be using `Sys.time()`... happy to make that a follow-on PR. It would give us a lot more certainty regarding uniqueness.